### ISSUE TITLE
Resolved Issues incorrectly displaying on Furniture records

### ISSUE THREAD

**Issue by am-esdm** [exeGesIS-SDM/CAMS-Mobile#76]
Issues recorded on Furniture in CAMS Mobile, then set to a Status of 'Resolved' (with the Resolved status update occurring on CAMS Mobile or in CAMS Desktop) continue to show as Issues on the associated Furniture record form in CAMS Mobile after a full sync-import-sync cycle via CAMS Mobile/Desktop/Mobile, and in the case of Point Furniture records, also as an attached Issue decoration symbology on the Furniture icon in the map but only after first viewing the Furniture form, setting the current Furniture condition to 'Good' (or another positive condition value that doesn't prompt a new Issue form) to allow you to then click to view the existing Issue form, then viewing the existing Issue form and returning to the map. Resolved Issues shouldn't display on CAMS Mobile after they have been synced back to CAMS Desktop, accepted and imported in CAMS Desktop and the result synced back to CAMS Mobile (or indeed after just syncing up resolved Issue data from CAMS Desktop to Mobile), and this has recently been sorted out for standalone Issue records on a parent Link record (see https://github.com/exeGesIS-SDM/CAMS-Mobile/issues/1) where resolved Issues on Links also used to persist instead of being removed. The same now needs to be applied to Issues on Furniture. Also see https://github.com/exeGesIS-SDM/CAMS-Mobile/issues/41 for a bit of loosely related further history in this area

**Comment by Timothy-D** [comment]
Observed this is occurring in 4.4.9

**Comment by LaineS** [comment]
The furniture map feature should show a small yellow triangle when:
* After syncing, the CAMS API sets the furniture.HasIssues field. If this is true, it will show a yellow icon. TD confirmed the API was setting this to true after checking if there were any unresolved issues on the furniture feature (i.e. so this should be ok)
* After the user edits an issue, the app now checks to see if a furniture map feature has issues and the status of each is 'resolved'. If there are no unresolved issues, the yellow triangle is removed

So hopefully this now works correctly.

**Comment by am-esdm** [comment]
**CAMS Mobile version:** v4.4.10
**Sync account:** xgWCP
**Tablet:** Samsung Galaxy Tab A
**Android version:** v5.1.1

**Result:**
Fail: 

There are two key areas mentioned in the task description above:

1. _Removal of resolved issue **records** from furniture after sync cycle back to CAMS Desktop (sync from CAMS Mobile, accept and process in to database from CAMS Desktop, re-sync accepted record on CAMS Mobile)._ After testing this: Fail: Resolved issues are still present within the furniture record form, after this cycle

2. _Removal of furniture issue **decoration** after issue on furniture is set to resolved_ - this is covered by https://github.com/exeGesIS-SDM/CAMS-Mobile/issues/23 where I've put testing feedback, copied here just for reference:

"Fail: The furniture's issue decoration is successfully removed when you initially save the record after updating the issue to resolved, but then if you view the issue form from the furniture again (and don't change anything in the issue record), then cancel back to the map, the issue decoration reappears for the now-resolved issue - it should stay removed"

**Comment by Timothy-D** [comment]
**Device**: Samsung Galaxy Tab Active 2
**OS Version**: Android 9
**App Version**: 4.4.10
**Sync Account**: xgWCP
**Status**: Fail
**Details**:
When a furniture item contains a resolved issue only and is synchronised down to the tablet, the symbol is drawn on the map without the /!\ decoration. This is OK.
However, if you view the details of the furniture item, the resolved issue is still shown on the furniture form. It shouldn't appear.
If you then go into the details of the issue, and cancel it, on returning to the map the furniture icon is shown with the /!\ decoration. (it shouldn't do - c.f. #23 )

**Comment by LaineS** [comment]
Discussed with TD. The furniture item should only show:
1. Issues that are not resolved
AND
2. Issues that are revolved but not yet uploaded

I think this is how the app currently works, so I've not made any changes here (NB #23 is fixed so the furniture decoration should be correct now)

**Comment by Timothy-D** [comment]
**Device**: Samsung Galaxy Tab Active 2
**OS Version**: Android 9
**App Version**: 4.4.11
**Sync Account**: xgWCP
**Status**: Fail
**Details**:
1. Find furniture item with no issue associated with it on the map in the mobile app.
2. Edit item, add issue and save.
3. Furniture item shows with /!\ decoration. OK
4. Go back into furniture item and issue is shown. OK
5. Sync data to CAMS and accept
6. Sync data in mobile app
7. Furniture item shows with /!\ decoration. OK
8. Go back into furniture item and issue is shown. OK
9. In CAMS Desktop, update issue to resolved.
10. Sync data in CAMS Mobile
11. Furniture item shows without /!\ decoration - OK
12. Open furniture details - issue is still listed there as resolved - NG

**Comment by Timothy-D** [comment]
**Device**: Samsung Galaxy Tab Active 2
**OS Version**: Android 9
**App Version**: 4.4.12
**Sync Account**: xgwcp
**Status**: Fail
**Details**:
1. Find furniture item with no issue associated with it on the map in the mobile app.
2. Edit item, add issue and save.
3. Furniture item shows with /!\ decoration. OK
4. Go back into furniture item and issue is shown. OK
5. Sync data to CAMS and accept
6. Sync data in mobile app
7. Furniture item shows with /!\ decoration. OK
8. Go back into furniture item and issue is _not_ shown. NG
9. In CAMS Desktop, update issue to resolved.
10. Sync data in CAMS Mobile
11. Furniture item shows without /!\ decoration - OK
12. Open furniture details - issue is still listed there as resolved - OK

Checking other furniture items on the map showing the /!\ decoration - none show issues on the furniture form.

1. Added a new issue to a furniture item in CAMS Desktop
2. Sync'd data from CAMS Mobile
3. Furniture item shows on map with /!\ decoration - OK
4. Open detail form for this furniture item. Issue is not displayed - NG

It appears that issues are not showing on furniture items at all, with the exception of newly created ones that have not been synchronised.

Display of the issue list on a furniture item should be as follows:
* Unresolved issues should always be shown.
* If an issue on a furniture item is resolved on the tablet, the issue should still display on the furniture form until the record has been synchronised & accepted in CAMS Desktop.
* Issues resolved elsewhere should not be shown.

**Comment by am-esdm** [comment]
> **Device**: Samsung Galaxy Tab Active 2
> **OS Version**: Android 9
> **App Version**: 4.4.12
> **Sync Account**: xgwcp
> **Status**: Fail
> **Details**:
> 
> 1. Find furniture item with no issue associated with it on the map in the mobile app.
> 2. Edit item, add issue and save.
> 3. Furniture item shows with /!\ decoration. OK
> 4. Go back into furniture item and issue is shown. OK
> 5. Sync data to CAMS and accept
> 6. Sync data in mobile app
> 7. Furniture item shows with /!\ decoration. OK
> 8. Go back into furniture item and issue is _not_ shown. NG
> 9. In CAMS Desktop, update issue to resolved.
> 10. Sync data in CAMS Mobile
> 11. Furniture item shows without /!\ decoration - OK
> 12. Open furniture details - issue is still listed there as resolved - OK
> 
> Checking other furniture items on the map showing the /!\ decoration - none show issues on the furniture form.
> 
> 1. Added a new issue to a furniture item in CAMS Desktop
> 2. Sync'd data from CAMS Mobile
> 3. Furniture item shows on map with /!\ decoration - OK
> 4. Open detail form for this furniture item. Issue is not displayed - NG
> 
> It appears that issues are not showing on furniture items at all, with the exception of newly created ones that have not been synchronised.
> 
> Display of the issue list on a furniture item should be as follows:
> 
> * Unresolved issues should always be shown.
> * If an issue on a furniture item is resolved on the tablet, the issue should still display on the furniture form until the record has been synchronised & accepted in CAMS Desktop.
> * Issues resolved elsewhere should not be shown.

Hi @Timothy-D please can you clarify this bit: "12. Open furniture details - issue is still listed there as resolved - OK" - it seems to contradict your bullet point '8', and also fully-accepted resolved issues shouldn't display on CAMS Mobile - so presumably the result of test step 12 should be fail (NG? (not sure what that acronym means)), not OK? Thanks

**Comment by Timothy-D** [comment]
Comment above updated (copy & paste mistake) - in 12 above, issue has been resolved and sync'd (9 & 10) so at this point as the issue is resolved, it shouldn't show
"NG" = "No Good"

**Comment by LaineS** [comment]
Hopefully should be fixed now

* Unresolved issues should always be shown.
* If an issue on a furniture item is resolved on the tablet, the issue should still display on the furniture form until the record has been synchronised & accepted in CAMS Desktop.
* Issues resolved elsewhere should not be shown.

**Comment by Timothy-D** [comment]
**Device**: Samsung Galaxy Tab Active 2
**OS Version**: Android 9
**App Version**: 4.4.13
**Sync Account**: xgwcp
**Status**:
**Details**:
Test 1
1. Find furniture item with no issue associated with it on the map in the mobile app.
2. Edit item, add issue (unresolved) and save.
3. Furniture item shows with /!\ decoration. OK
4. Go back into furniture item and issue is shown. OK
5. Sync data to CAMS but don't process
6. Check map: Furniture item shows with /!\ decoration. OK
7. Go back into furniture item and issue is shown. OK
8. Process data in CAMS
9. Sync data on tablet
10. Check map: Furniture item shows with /!\ decoration. OK
11. Go back into furniture item and issue is shown. OK
12. On tablet, go to issue and update to resolved.
13. Furniture icon shows without the /!\ decoration. OK
14. Check record: issue shows against the furniture item. OK
15. Sync data to CAMS but don't process
16. Furniture Map icon shows with the /!\ decoration. Q?
17. Check record: issue shows against the furniture item, showing as unresolved. NG
18. Process data in CAMS
19. Sync data on tablet
20. Furniture icon shows without the /!\ decoration. OK
21. Check record: no issue shown. OK

Test 2
1. Find furniture item with no issue associated with it on the map in the mobile app.
2. Edit item, add issue (unresolved) and save.
3. Furniture item shows with /!\ decoration. OK
4. Go back into furniture item and issue is shown. OK
5. Sync data to CAMS but don't process
6. Check map: Furniture item shows with /!\ decoration. OK
7. Go back into furniture item and issue is shown. OK
8. Process into CAMS
9. Update issue to resolved in CAMS Desktop
10. Sync data on tablet
11. Furniture icon shows without the /!\ decoration. OK
12. Check record: no issue shown. OK

**Comment by Timothy-D** [comment]
Repeated Test1 above on a different furniture item, with the same results. This seems consistent.

**Comment by LaineS** [comment]
There was a bug in the logic and I've switched that to an AND (it was OR). This should be fixed now in v4.4.14. 

However, there is a larger problem: 

1. Create an issue that is NOT resolved on a furniture item. The ! will show on the map (OK)
2. Sync data and process it in CAMS Desktop. Re-sync to the device, the ! is still shown (OK)
3. Resolve the issue on the mobile app and save. The ! is removed (OK)
4. Sync with the API, but do not process in CAMS. The ! will be shown on the furniture item (NOT OK)

This is because when the mobile app gets data from CAMS API, it sends furniture.hasIssues = true back to the mobile device (because as far as CAMS Desktop is concerned there is still an unresolved issue on the furniture item, the update is still pending in the holding table and has not been processed). 

So the mobile app sets the furniture ! icon on the feature. I'm not sure how to fix this, can you review and discuss what you would like to happen?

(NB I am on leave now until 4th Jan)

**Comment by Timothy-D** [comment]
1. Find furniture item with no issue associated with it on the map in the mobile app.
2. Edit item, add issue (unresolved) and save.
3. Furniture item shows with /!\ decoration. OK
4. Go back into furniture item and issue is shown. OK
5. Sync data to CAMS but don't process
6. Check map: Furniture item shows with /!\ decoration. OK
7. Go back into furniture item. Issue is _not_ shown. Q?
8. Process data in CAMS
9. Sync data on tablet
10. Check map: Furniture item shows with /!\ decoration. OK
11. Go back into furniture item and issue is shown. OK
12. On tablet, go to issue and update to resolved.
13. Furniture icon shows without the /!\ decoration. OK
14. Check record: issue shows against the furniture item. OK
15. Sync data to CAMS but don't process
16. Furniture Map icon shows with the /!\ decoration. Q?
17. Check record: issue shows against the furniture item, showing as unresolved. NG
18. Process data in CAMS
19. Sync data on tablet
20. Furniture icon shows without the /!\ decoration. OK
21. Check record: no issue shown. OK

**Comment by Timothy-D** [comment]
In database, need to check after points 4, 7, 11, 17, 20:
Infra.HasIssues
Infra.IsSubmitted
Maint.Status
Maint.IsSubmitted

**Comment by Timothy-D** [comment]
1. Find furniture item with no issue associated with it on the map in the mobile app.
1. Edit item, add issue (unresolved) and save.
1. Furniture item shows with /!\ decoration. OK
1. Go back into furniture item and issue is shown. OK
1. Sync data to CAMS but don't process
1. Check map: Furniture item shows with /!\ decoration. OK
1. Go back into furniture item. Issue is not shown. Q?
1. Process data in CAMS
1. Sync data on tablet
1. Check map: Furniture item shows with /!\ decoration. OK
1. Go back into furniture item and issue is shown. OK
1. On tablet, go to issue and update to resolved.
1. Furniture icon shows without the /!\ decoration. OK
1. Check record: issue shows against the furniture item. OK
1. Sync data to CAMS but don't process
1. Furniture Map icon shows with the /!\ decoration. Q?
1. Check record: issue shows against the furniture item, showing as unresolved. NG
1. Process data in CAMS
1. Sync data on tablet
1. Furniture icon shows without the /!\ decoration. OK
1. Check record: no issue shown. OK

G_ISSUE_STATUS_RESOLVED_VALUE=1

| Step | InfraCode | Infra GUID | Infra.IsSubmitted | Infra.HasIssues |  MaintCode | Maint GUID | Maint.IsSubmitted | Maint.Status | Decoration | Issue | Overall |
| --- | --- | --- | --- | --- | --- | --- | --- | --- | --- | --- | --- |
| 4 | Gwyne/62 | 4940dcc2-77e4-4993-8ec2-ca957809789b | for_update | true |  | b47988ee-223a-4937-a742-7ac119a642a9 | for_insert | 3 | Yes | Yes | OK |
| 5 | Gwyne/62 | 4940dcc2-77e4-4993-8ec2-ca957809789b | submitted for_update | true |  | b47988ee-223a-4937-a742-7ac119a642a9 | submitted for_insert | 3 | Yes | No | Qry |
| 9 | Gwyne/62 | 4940dcc2-77e4-4993-8ec2-ca957809789b |  | true | 07678 | 567d161e-ec25-4cfc-8865-7ee380662bec |  | 3 | Yes | Yes | OK |
| 14 | Gwyne/62 | 4940dcc2-77e4-4993-8ec2-ca957809789b | for_update | 0 | 07678 | 567d161e-ec25-4cfc-8865-7ee380662bec | for_update | 1 | No | Yes | OK |
| 15 | Gwyne/62 | 4940dcc2-77e4-4993-8ec2-ca957809789b |  | true | 07678 | d0e58c91-e78b-4c6e-83d1-05e7771b75f3 |  | 3 | Yes | Yes - as unresolved | NG |
| 19 | Gwyne/62 | 4940dcc2-77e4-4993-8ec2-ca957809789b |  | false | 07678 | 9362c5cc-5c06-4ec7-8491-e8421e3056df |  | 1 | No | No | OK |

Where teh data are out of kilter, this seems to be when the record has been synced up but not processed.

Queries:
* Why does the Maint GUID keep changing?
* The sync at step 15 seems to update the data differently to the sync at step 5. Why is this?
* What sets the issue status back to unresolved during the sync at step 15?

**Comment by am-esdm** [comment]
Hi @LaineS, just caught up with TD on this one, would you mind providing a response to his most recent queries above please? That will inform next steps. Thanks

**Comment by LaineS** [comment]
We will need to discuss this via Teams I think

**Comment by Timothy-D** [comment]
This will need re-testing to confirm the current situation before anything can be implemented.

**Comment by Timothy-D** [comment]
select furniture.infracode, furniture.guid as InfraGuid, furniture.IsSubmitted as InfraSubmitted, furniture.HasIssues, issue.MaintCode, issue.guid as MaintGuid, issue.IsSubmitted as MaintSubmitted, issue.status
from furniture join issue on furniture.infracode = issue.ParentCode
where infracode = '{InfraCode}'

**Comment by Timothy-D** [comment]
**Device**: Emulator
**OS Version**: Android 8
**App Version**: 4.5.0
**Sync Account**: xgdev
**Details**:
The table below shows the different stages in the workflow, after each step, the furniture icon on the map was checked for the issue decoration, and the furniture form checked for the presence of the issue in the issue grid. After each step, the database was exported and values checked in there.

| Step | Operation | Decoration Expected | Decoration shown | Issue expected | issue shown | Overall | InfraCode | InfraGUID | Furniture.IsSubmitted | Furniture.HasIssues | maintCode | MaintGUID | Issue.IsSubmitted | Issue.Status |
| --- | --- | --- | --- | --- | --- | --- | --- | --- | --- | --- | --- | --- | --- | --- |
| 1 | Find furniture item with no issue associated with it on the map in the mobile app.  | No | No | No | No | OK | 006/24 | 035fc61c-27ad-40ee-9d96-344d5e61f77d | {null} | false | {null} | {null} | {null} | {null} |
| 2 | Edit Furniture item set condition to unsatisfactory, add issue (unresolved) and save both. | Yes | Yes | Yes | Yes | OK | 006/24 | 035fc61c-27ad-40ee-9d96-344d5e61f77d | for_update | 1 | {null} | ae72d2db-9578-4b57-9e9d-527ae72feda1 | for_insert | 3 |
| 3 | Sync data to CAMS but don't process | Yes | Yes | Yes | No | NG | 006/24 | 035fc61c-27ad-40ee-9d96-344d5e61f77d | submitted for_update | true | {null} | ae72d2db-9578-4b57-9e9d-527ae72feda1 | submitted for_insert | 3 |
| 4 | Process data in CAMS. Sync data on tablet | Yes | Yes | Yes | Yes | OK | 006/24 | 28df6d00-8c39-43d4-8492-afa7dbffe653 | {null} | true | 00448 | a8c4aec9-17a4-4eb1-8982-d840a8e04e18 | {null} | 3 |
| 5 | On tablet, go to furniture item & set to satisfactory, update issue to resolved. Save changes to issue & furniture. | No | No | Yes - as resolved | Yes - as resolved | OK | 006/24 | 28df6d00-8c39-43d4-8492-afa7dbffe653 | for_update | false | 00448 | a8c4aec9-17a4-4eb1-8982-d840a8e04e18 | for_update | 1 |
| 6 | Sync data to CAMS but don't process | No | Yes | No | Yes - as Unresolved | NG | 006/24 | 7762af00-96c4-4172-a7a1-df285f21f658 | {null} | true | 00448 | b0d4675a-5b82-46f5-a572-f839a6b55860 | {null} | 3 |
| 7 | Process data in CAMS. Sync data on tablet | No | No | No | No | OK | 006/24 | fe22fd7e-cbaf-47cb-b25f-3d2b7b807c76 | {null} | false | 00448 | 54445a8a-8d2d-41d8-a286-1decef7fe3c9 | {null} | 1 |

Main issue seems to be when data are uploaded but have not been processed and downloaded again. In particular at step 6 above, the issue is being set back to unresolved. Why is this happening?

@KathrynSteemson need to discuss this one.

**Comment by Timothy-D** [comment]
Questions for @LaineS 
- Why is the issue getting "reset" at steps 3 & 6 above?
- Is there an easy way that the local updates can remain on the tablet, or would this have other (undesirable) consequences?

**Comment by LaineS** [comment]
I'm not entirely sure, the logic is very convoluted around what is going on. If things are working ok, I would suggest leaving them as is.
